rfd/0205-improved-onprem-joining.md

```diff
@@ -0,0 +1,1010 @@
+
+   If they do not match, the join is rejected, and a lock is generated against
+   the affected `(bot, token)`.
+
+Just as with the generation counter, this procedure relies on an imposter bot
+successfully joining once. When the original bot fails to refresh and attempts
+to rejoin, it presents a valid but outdated join state document, we generate a
+lock, and then deny further access to both bots.
+
+#### Token Resource Example
+
+`bound-keypair`-type tokens differ from other types in that they are intended to
+have no resource-level expiration (though that is allowed), are meant to have
+their spec modified over time by users or automation tools, and publish
+information about their current state in the immutable (to users) `status`
+field.
+
+``` yaml
+kind: token
+version: v2
+metadata:
+  name: my-join-token
+spec:
+  bot_name: example
+
+  join_method: bound-keypair
+  bound_keypair:
+    # `onboarding` parameters control initial join behavior
+    onboarding:
+      # If set, no joining secret is generated; the secret exchange ceremony is
+      # skipped and instance will directly prove its identity using its private
+      # key. It is an error for a public key to be associated with more than one
+      # token, and creation or update will fail if a public key is reused.
+      # May not be modified after resource creation. Note that public keys may
+      # be rotated, so refer to `.status.bound_keypair.bound_public_key` for the
+      # currently bound key information.
+      initial_public_key: null
+
+      # If set, use an explicit initial joining secret; if both this and
+      # `public_key` are unset, a value will be generated server-side and
+      # stored in `.status.bound_keypair.initial_join_secret`
+      initial_join_secret: ""
+
+      # Initial joining must take place before this timestamp. May be
+      # modified if bot has not yet joined.
+      must_join_before: "2025-03-01T21:45:40.104524Z"
+
+    # Parameters to tune rejoining behavior when the regular bot identity has
+    # expired
+    rejoining:
+      # If true, `total_rejoins` is ignored and bots may rejoin indefinitely;
+      # must be opt-in.
+      unlimited: false
+
+      # Total number of allowed rejoins; this may be incremented to allow
+      # additional rejoins, even if a bot identity has already expired. May
+      # be decremented, but only by the current value of
+      # `.status.bound_keypair.remaining_rejoins`.
+      total_rejoins: 10
+
+      # If set, rejoining is only valid before this timestamp; may be
+      # incremented or reset to the empty string to allow rejoining once
+      # expired.
+      may_rejoin_until: "2026-03-01T21:45:40.104524Z"
+
+    # If set, the bot will perform a keypair rotation on its next renewal after
+    # it is informed of the change to this field. Note that this is tied to bot
+    # heartbeats and may not take effect on the next refresh interval.
+    rotate_on_next_renewal: false
```

Comment:
Does this field get automatically reset when the bot rotates its public key?

I'm wondering if it's possible for an admin to set this and forget about it (leading to rotations on every renewal). Which arguably wouldn't be the worst thing in the world, just a little wasteful.

Reply:
Writing down what we discussed, yes, this field should be cleared after a rotation. It's alluded to in the flowchart, but I'll update the doc to make that clear.